### ISSUE TITLE
Review comments

### ISSUE THREAD

**Issue by kellertuer** [PdIPS/CBX#3]
Thanks for the nice paper and for submitting it to JOSS (cf https://github.com/openjournals/joss-reviews/issues/6611) . Here are a few comments both on the content of the rendered PDF of this repositories paper as well as on the Julia package. I will do them as check boxes but also number them in case you would like to comment / discuss. Once we agree they are finished, I can check them then.

## The paper

Line numbers in the following points refer to the version generated by the editorialbot on April 12, 2024 [here](https://github.com/PdIPS/CBX/issues/[https://raw.githubusercontent.com/openjournals/joss-papers/joss.06611/joss.06611/10.21105.joss.06611.pdf) 
* [ ] 1. line 32ff. The authors have a nicve motivation and overview. For a unified presentation one could rewrite “where $\nabla f$ is not well-defined” also more like the part before “where f is nonsmooth ($\nabla f$ is not well-defined),or where the evaluation […]”
* [ ] 2. line 40: One could maybe mention that $t$ is time already in the sentence defining $x_t$ instead of in the next – but this might also still be fine as is.
* [ ] 3. line 99/100: a small remark for the statement of need. The authors state Optimization.jl also includes Nelder Mead. The other mentioned package (Manopt.jl) also includes Nelder-Mead and most recently (probably after submission of the authors) also CMA-ES, for which I am not sure how much this is related to the CBX family of solvers.
* [ ] 4. line 105: “easy-to-understand” is subjective, maybe one could just leave that out? I feel “high-level” is enough of an explanation that even beginner users should get an easy start.
* [ ] 5. It would maybe be nice – for example after the specifics – to do a short tutorial or a small example to illustrate how to use the package? If could also be within the “specifics”, then I would leave our that word, and add for both packages a small code snipped others could get started with. Since the Markdown is available, readers can even copy them, but a 5 line example or so would be nice to get started

## The Julia package

* [ ] 6. A contributing.md would maybe be nice explaining a bit how other users could contribute to the package.
* [ ] 7. I personally feel the aliases minimize/minimise/optimise/optimize are more confusing than helpful 
* [ ] 8. The “Full code example” links in the documentation could also directly be put into the docs instead of just linked. For example as plain code blocks, but maybe even as Literate.jl or Quarto-rendered ones so the documentation also reports on the results.
* [ ] 9. Concerning tests – you have a quite high code coverage already – You could consider even reaching for a bit more, since it self the configuration calls can easily be tested; are for example the errors actually ever hit?. As for the parts in `utils/arrays.jl` maybe those are not necessary as functions when they are never called? In run-dynamics there is a few return nothing lines that seem to be never reached, are they necessary then?
* [x] 10. Since this is more a comment than a request, I already checked this one. Since you have quite a few citations in the docs, I felt it was great to use https://github.com/JuliaDocs/DocumenterCitations.jl in a documentation to have a way to cite papers.

**Comment by TimRoith** [comment]
Thank you very much for the comments! Since I am mainly concerned with the Python implementation, I will first only address the issues regarding the paper:

* Ad 1. We unified the presentation accordingly, and mimic the bracket structure from before.
* Ad 2. We introduce $t$ in the first sentence now.
* Ad 3. We now mention CMA-ES, especially we cite the Python and Julia implementations for it. The connection between CMA-ES and consensus based methods, is indeed interesting, however to the best of our knowledge, unexplored. In fact, it is part of ongoing work, therefore it is really important not forgetting to mention it here.
*Ad 4. We removed "easy-to-understand".
* Ad 5. For Python a small example code snippet is included, for Julia a similar one will follow.

**Comment by rafaelbailo** [comment]
Thank you for the comments @kellertuer! I believe points 5 and 6 are now addressed.

Regarding 7: with the exception of the note about the aliases, the documentation for CBO is written entirely in terms of `minimise`. The aliases can of course be ignored, but are there to increase compatibility with users from other communities: I have colleagues who always "minimise", and I have colleagues who always "optimise", and both groups can be quite vocal about their terminology. If it's alright with you, I would prefer to keep the aliases, just to avoid future headaches!

**Comment by rafaelbailo** [comment]
Regarding 8: we considered using Literate.jl for the code examples to make the documentation easier to manage. However, we found that the examples are quite static (since the top-level interface is quite simple and very unlikely to change), and simultaneously we wanted to use the examples as part of the unit testing: https://github.com/PdIPS/ConsensusBasedX.jl/blob/main/test/examples.jl checks that every example can be run and generates no errors or warnings. Is there a good way to merge the testing and something like Literate.jl?

**Comment by rafaelbailo** [comment]
Regarding 9: I've just pruned the superfluous methods and added a couple of additional unit tests. Our (free) codecov rate has maxed, but I will upload the new coverage report ASAP!

**Comment by kellertuer** [comment]
> I have colleagues who always "minimise", and I have colleagues who always "optimise", and both groups can be quite vocal about their terminology. 

I minimise by optimisation. ;)
I think my main complaint is really about _both_ American _and_ British English aliases. even without that I feel a package could have a consistent (that is one) terminology and other names could be mentioned in the docs.

**Comment by kellertuer** [comment]
Concerning 8, one could also include the 3-4 lines just as code in the docs instead of putting them a click away.
I personally use Quarto, but Literate or Weave are not _too_ far way from that; still, avoiding a click to the example would already be the feature that would make me happy.

The remaining comments I will check the next days

**Comment by gdalle** [comment]
I had not noticed this issue when I opened my own review issue here: https://github.com/PdIPS/ConsensusBasedX.jl/issues/31

Some of the concerns of @kellertuer are shared, so I'll jump in here.

>  I think my main complaint is really about both American and British English aliases. even without that I feel a package could have a consistent (that is one) terminology and other names could be mentioned in the docs.

Definitely agree with this one. Picking a side is not a bad thing.

> Concerning 8, one could also include the 3-4 lines just as code in the docs instead of putting them a click away.

Also a major agreement on my end.

>    > Is there a good way to merge the testing and something like Literate.jl?
>
> I am not aware of something doing that.

Not automatically, but you can use the same source files and include them both in docs and in tests. My package [HiddenMarkovModels.jl](https://github.com/gdalle/HiddenMarkovModels.jl) does that, although it might be slightly more complicated when you have several source files per docs page.

> https://github.com/PdIPS/ConsensusBasedX.jl/blob/main/test/examples.jl checks that every example can be run and generates no errors or warnings.

Are the examples ever checked for correctness? More generally, my [other major issue](https://github.com/PdIPS/ConsensusBasedX.jl/issues/31#issuecomment-2094108786) was that the tests never seem to verify that the result is the minimiser we expect

**Comment by gdalle** [comment]
It is much nicer for users, thank you!
Do the examples ever run during docs build or tests? Is there anything preventing them from getting outdated without you noticing?

**Comment by rafaelbailo** [comment]
Yes, the unit tests still run every single file in the examples folder, checking that the code executes without errors.

**Comment by gdalle** [comment]
But is the result ever checked against what is expected? We can maybe discuss it over in my review since it was my main worry, don't want to hijack this one: https://github.com/PdIPS/ConsensusBasedX.jl/issues/31#issuecomment-2094108786

**Comment by kellertuer** [comment]
What did you do to the optimise/minimize comment that is still an alias?

**Comment by rafaelbailo** [comment]
Sorry, I understood that the problem was the British/American alias:
> I minimise by optimisation. ;) I think my main complaint is really about _both_ American _and_ British English aliases. even without that I feel a package could have a consistent (that is one) terminology and other names could be mentioned in the docs.

Shall I remove `optimise` also?

**Comment by kellertuer** [comment]
No, the problem was actually both.

I think a package should have exactly one language they use. Sure the documentation can state “People form this area might refer to A as B”. But if you provide 2 functions, why not also the German, French, Norwegian, Swedish,... translations?
So with “language” I meant both the different language aliases _as well as_ the different terminology aliases.

So please decide to use only one of those 4.

**Comment by rafaelbailo** [comment]
Alright, I've removed `optimise` 👍🏻

**Comment by kellertuer** [comment]
Wrong choice. I should of course have even `optimise` to keep ;) No I am joking, that would just be _my_ terminology, but I am in no place to prescribe you a terminology.

Good that you narrowed it down to one.
